app/models/training_period.rb

```diff
@@ -41,6 +41,10 @@ class TrainingPeriod < ApplicationRecord
   validate :trainee_distinct_period
   validate :enveloped_by_trainee_at_school_period
   validate :only_provider_led_mentor_training
+  validates :api_withdrawn_at, presence: true, if: -> { api_withdrawal_reason.present? }
+  validates :api_withdrawal_reason, presence: true, if: -> { api_withdrawn_at.present? }
+  validates :api_deferred_at, presence: true, if: -> { api_deferral_reason.present? }
+  validates :api_deferral_reason, presence: true, if: -> { api_deferred_at.present? }
```

Comment:
I know this is in draft so just an initial though, but I assume someone's training can either be withdrawn or deferred, not both. In some of the earlier discussions we had this as `closure_reason` with a single enum that had a prefix, so `withdrawn-left-teaching-profession` or `deferred-bereavement`. Then there's no need to add validation to make sure there's only one present.

Reply:
So I think you're right in that a `TrainingPeriod` could only be withdrawn or deferred but not both. The only knock-on effect I can see with combining them is that we would need to format the value when we accept/return it from the API, so map to/from:

`withdraw-left-teaching-profession`/`left-teaching-profession`

That shouldn't be a big deal though and we do it already for withdrawal reasons after a change to some of the programme type namings (see [this in ECF](https://github.com/DFE-Digital/early-careers-framework/blob/main/app/services/programme_type_mappings.rb#L9-L18)). I'm going to hold off for Abeer's thoughts before I rework it though in case I'm missing anything; I'll add the mutually exclusive validation for now to the existing attributes 👍

Reply:
Ah, yeah - that makes sense!

Reply:
Hey @peteryates, I caught up with @cwrw this morning and we discussed the idea of combining the timestamp and reason fields. 

It feels like combining them would make it clearer that there’s only one reason to close off a training period and would reduce validation. However, it will also complicate API slightly with mapping to/from and it feels like it makes the overall deferred/withdrawn state less clear. It will also add a bit of complexity to querying and conditional logic (since we’d have to inspect the reason to determine state). For these reasons, we think it’s going to make more sense to keep them separate. Does that sound reasonable?